File: android/app/src/main/java/kg/ayant/app/core/AppConfig.kt
package kg.ayant.app.core
import kg.ayant.app.data.FirebasePointsRepository
import kg.ayant.app.data.MockPointsRepository
import kg.ayant.app.data.FirebaseFileUploadService
import kg.ayant.app.data.MockFileUploadService
import kg.ayant.app.data.FirebasePushService
import kg.ayant.app.data.MockPushService
import kg.ayant.app.data.MockDataRepository

import android.content.Context
import kg.ayant.app.domain.contract.AnalyticsService
import kg.ayant.app.domain.contract.AuthService
import kg.ayant.app.domain.contract.CouponService
import kg.ayant.app.domain.DataRepository
import kg.ayant.app.data.DemoAnalyticsService
import kg.ayant.app.data.FirebaseAnalyticsService
import kg.ayant.app.data.FirebaseAuthService
import kg.ayant.app.data.FirebaseCouponService
import kg.ayant.app.data.FirebaseDataRepository
import kg.ayant.app.data.FirebaseRankingEventService
import kg.ayant.app.data.MockAnalyticsService
import kg.ayant.app.data.MockAuthService
import kg.ayant.app.data.MockCouponService
import kg.ayant.app.data.MockRankingEventService
import kg.ayant.app.domain.contract.RankingEventService

/**
 * Single switch between Mock and Firebase, mirroring AppConfig.swift.
 *
 * To go live on the shared Firebase project (san-25d32):
 *  1. In the Firebase console, add an Android app with package kg.ayant.app.
 *  2. Download google-services.json into android/app/.
 *  3. Flip useFirebase to true and uncomment the Firebase branch below +
 *     the deps in app/build.gradle.kts (already present).
 */
object AppConfig {
    var useFirebase = false

    lateinit var applicationContext: Context

    fun makeDataRepository(): DataRepository =
        if (useFirebase) FirebaseDataRepository() else MockDataRepository()

    fun makeAuthService(): AuthService =
        if (useFirebase) FirebaseAuthService(applicationContext) else MockAuthService()

    /**
     * ВРЕМЕННО: показывать в «Аналитике» сгенерированные ряды вместо реальных.
     *
     * События при этом продолжают писаться в настоящий сервис, поэтому
     * выключение флага вернёт накопленные данные, а не пустоту.
     */
    const val useDemoAnalytics = true

    fun makeAnalyticsService(): AnalyticsService {
        val real = if (useFirebase) FirebaseAnalyticsService() else MockAnalyticsService()
        return if (useDemoAnalytics) DemoAnalyticsService(real) else real
    }

    fun makeRankingEventService(): RankingEventService =
        if (useFirebase) FirebaseRankingEventService() else MockRankingEventService()

    /** Push-темы и регистрация токена. Вне Firebase — заглушка. */
    fun makePushService(): kg.ayant.app.domain.contract.PushService =
        if (useFirebase) kg.ayant.app.data.FirebasePushService()
        else kg.ayant.app.data.MockPushService()

    /** Загрузка файлов (фото заведений, PDF-меню). Вне Firebase — заглушка. */
    fun makeFileUploadService(): kg.ayant.app.domain.contract.FileUploadService =
        if (useFirebase) kg.ayant.app.data.FirebaseFileUploadService()
        else kg.ayant.app.data.MockFileUploadService()

    /** Живой источник карт баллов (snapshot-листенер вместо опроса) + списание. */
    fun makePointsRepository(): kg.ayant.app.domain.PointsRepository =
        if (useFirebase) kg.ayant.app.data.FirebasePointsRepository(makeCouponService(), makeAuthService())
        else kg.ayant.app.data.MockPointsRepository(kg.ayant.app.domain.MockData.pointsCards)

    fun makeCouponService(): CouponService =
        if (useFirebase) FirebaseCouponService() else MockCouponService()
}

File: SAN/AppConfig.swift
import Foundation
import AyantDomain
import AyantData

/// Единая точка переключения между mock-реализацией и Firebase.
///
/// Когда будешь готов подключить Firebase:
/// 1. Добавь пакет firebase-ios-sdk (SPM) и GoogleService-Info.plist.
/// 2. Раскомментируй Firebase-ветки ниже и в FirebaseServices.swift.
/// 3. Поставь useFirebase = true.
enum AppConfig {

    /// Переключатель mock ⇄ Firebase. Ставь true, когда настроишь консоль/ключи.
    static let useFirebase = true

    /// Базовый URL Cloud Functions — одно место вместо разбросанных по коду ссылок.
    /// При смене региона/проекта/бэкенда правится только здесь.
    /// Адрес живёт в слое данных (`AyantBackend`); здесь — привычная точка вызова.
    static let functionsBaseURL = AyantBackend.functionsBaseURL
    static func functionURL(_ name: String) -> String { AyantBackend.functionURL(name) }

    static func makeAuthService() -> AuthService {
        useFirebase ? FirebaseAuthService() : MockAuthService()
    }

    static func makeDataRepository() -> DataRepository {
        useFirebase ? FirebaseDataRepository() : MockDataRepository()
    }

    static func makeHostRepository() -> HostRepository {
        useFirebase ? FirebaseHostRepository() : MockHostRepository()
    }

    /// ВРЕМЕННО: показывать в «Аналитике» сгенерированные ряды вместо реальных.
    ///
    /// Пока у заведений нет трафика, настоящие счётчики — нули, и экран
    /// выглядит сломанным. События при этом продолжают писаться в настоящий
    /// сервис, поэтому выключение флага вернёт накопленные данные, а не пустоту.
    static let useDemoAnalytics = true

    static func makeAnalyticsService() -> AnalyticsService {
        let real: AnalyticsService = useFirebase ? FirebaseAnalyticsService() : MockAnalyticsService()
        return useDemoAnalytics ? DemoAnalyticsService(wrapping: real) : real
    }

    static func makeRankingEventService() -> RankingEventService {
        useFirebase ? FirebaseRankingEventService() : MockRankingEventService()
    }

    static func makePushService() -> PushService {
        useFirebase ? FirebasePushService() : MockPushService()
    }

    static func makeCouponService() -> CouponService {
        useFirebase ? FirebaseCouponService() : MockCouponService()
    }

    /// Продуктовая аналитика (DAU/воронки). Вне Firebase — печать в консоль.
    static func makeProductAnalytics() -> ProductAnalytics {
        useFirebase ? FirebaseProductAnalytics() : ConsoleProductAnalytics()
    }

    /// Живой источник карт баллов (snapshot-листенер вместо опроса) + списание.
    static func makePointsRepository() -> PointsRepository {
        useFirebase
            ? FirebasePointsRepository(backend: makeCouponService(), auth: makeAuthService())
            : MockPointsRepository(cards: MockData.pointsCards)
    }
}
